### PR TITLE
msbuild workspace global property defaults can be overridden

### LINKED ISSUE #33047: ArgumentException when trying to set property from ProjectBuildManager.s_defaultGlobalProperties
This is a follow-up to #30174 where the advice is to set `ShouldUnsetParentConfigurationAndPlatform` to `false` when building in `Release` configuration.

On the other hand when I try to set this property in latest version (3.0.0-beta2-final) then ArgumentException is thrown.

Does it means that properties from `ProjectBuildManager.s_defaultGlobalProperties` are not meant to be set? If this is the case then it would be better to provide more meaningful exception message (and maybe throw earlier). If the properties can be set then `AddRange` cannot be used.

https://github.com/dotnet/roslyn/blob/c8e0e7f29853eb5775da2187991e68fa5077f7af/src/Workspaces/Core/MSBuild/MSBuild/Build/ProjectBuildManager.cs#L153

**Steps to Reproduce**:
```
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.Build.Locator;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.MSBuild;
using System;

static class Program
{
    static async Task Main()
    {
        MSBuildLocator.RegisterDefaults();

        var properties = new Dictionary<string, string>
        {
            ["ShouldUnsetParentConfigurationAndPlatform"] = "false"
        };

        using (MSBuildWorkspace workspace = MSBuildWorkspace.Create(properties))
        {
            Solution solution = await workspace.OpenSolutionAsync(@"Foo.sln");
        }
    }
}

```

**Actual Behavior**:
```
System.ArgumentException
  HResult=0x80070057
  Message=An element with the same key but a different value already exists. Key: ShouldUnsetParentConfigurationAndPlatform
  Source=System.Collections.Immutable
  StackTrace:
   at System.Collections.Immutable.ImmutableDictionary`2.HashBucket.Add(TKey key, TValue value, IEqualityComparer`1 keyOnlyComparer, IEqualityComparer`1 valueComparer, KeyCollisionBehavior behavior, OperationResult& result)
   at System.Collections.Immutable.ImmutableDictionary`2.AddRange(IEnumerable`1 items, MutationInput origin, KeyCollisionBehavior collisionBehavior)
   at System.Collections.Immutable.ImmutableDictionary`2.AddRange(IEnumerable`1 pairs, Boolean avoidToHashMap)
   at System.Collections.Immutable.ImmutableDictionary`2.AddRange(IEnumerable`1 pairs)
   at Microsoft.CodeAnalysis.MSBuild.Build.ProjectBuildManager.StartBatchBuild(IDictionary`2 globalProperties)
   at Microsoft.CodeAnalysis.MSBuild.MSBuildProjectLoader.Worker.<LoadAsync>d__18.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.MSBuild.MSBuildProjectLoader.<LoadSolutionInfoAsync>d__21.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.<OpenSolutionAsync>d__23.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at MSBuildTest.Program.<OpenSolutionAndCompileAsync>d__1.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at MSBuildTest.Program.<Main>d__0.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at MSBuildTest.Program.<Main>(String[] args)
```

### PR DESCRIPTION
**Pull request by grahamdyson**
fixes #33047

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=46884) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=46884) before we can accept your contribution.<br/><br/>:x: grahamdyson [sign now](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=46884)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/roslyn?pullRequest=46884) it.</sub>

**Comment by grahamdyson** [comment]
looks like there might be an non-deterministic test unrelated to the changes
https://github.com/dotnet/roslyn/issues/46681

**Comment by jmarolf** [comment]
@grahamdyson re-queueing as I think this is a flaky test.

File: src/Workspaces/MSBuildTest/MSBuildWorkspaceTests.cs [hunk 1]
```diff
@@ -419,6 +419,46 @@ public async Task TestOpenSolution_LoadMetadataForReferencedProjects()
             }
         }
 
+        [ConditionalFact(typeof(VisualStudioMSBuildInstalled)), Trait(Traits.Feature, Traits.Features.MSBuildWorkspace)]
+        [WorkItem(33047, "https://github.com/dotnet/roslyn/issues/33047")]
+        public async Task TestOpenProject_CSharp_GlobalPropertyDesignTimeBuildDefault()
```
Comment (jmarolf) [inline src/Workspaces/MSBuildTest/MSBuildWorkspaceTests.cs:424]: Can you add a test for passing in a properties dictionary per https://github.com/dotnet/roslyn/issues/33047?

Comment (grahamdyson) [reply src/Workspaces/MSBuildTest/MSBuildWorkspaceTests.cs:424]: the other new test below passes in a properties dictionary, but with a non-default value for DesignTimeBuild and not ShouldUnsetParentConfigurationAndPlatform (as per #33047). both properties are specified in ProjectBuildManager.s_defaultGlobalProperties. i could change the test from DesignTimeBuild to ShouldUnsetParentConfigurationAndPlatform to exactly match the issue, but probably not worth in testing both?

**Comment by grahamdyson** [comment]
looks like the latest build failure might be due to #47122

**Comment by jmarolf** [comment]
Thanks so much @grahamdyson!

**Comment by grahamdyson** [comment]
> Thanks so much @grahamdyson!

thanks for everyone's help getting the fix in. i'm guessing everyone's got a lot on finishing .NET 5 😅

**Pull request merged**